### ISSUE TITLE
[BUG] Collection of Lines does not return correct field

### ISSUE THREAD

**Issue by hotteock** [magpylib/magpylib#368]
**Problem description**

Hello,

I have some question about Collection.
I want to draw Square-Coils(>=2), and then Calculate that Square-Coils Effect.

So, I used magpylib.current.Line() and Collection method.
Then, I calculated magnetic by Collection.getB(). I expected several square coils effect, but I just got one-coil effect.

here is my code

```
import magpylib as magpy
from magpylib import current
import numpy as np
import matplotlib.pyplot as plt

width = 2300
height = 2300
I = 60

vertex = [(-width/2.0, height/2.0, 0),
(width/2.0, height/2.0, 0),
(width/2.0, -height/2.0, 0),
(-width/2.0, -height/2.0, 0),
(-width/2.0, height/2.0, 0)]

coilLoop = magpy.current.Line(current=I, vertices=vertex, position=[0,0,-1000])
#coilLoop = magpy.current.Circular(current=I, diameter=1000, position=[0,0,-1000])
coilLoop2 = magpy.current.Line(current=I, vertices=vertex, position=[0,0,1000])
#coilLoop2 = magpy.current.Circular(current=I, diameter=1000, position=[0,0,1000])

plotBounds = [-3000, 3000, -3000, 3000]
nPlotPoints = 20

xs = np.linspace(plotBounds[0],plotBounds[1],nPlotPoints)
zs = np.linspace(plotBounds[2],plotBounds[3],nPlotPoints)
    
X,Z = np.meshgrid(xs,zs)



col = magpy.Collection()
col.add(coilLoop)
col.add(coilLoop2)

B = np.array([[col.getB([xs[x],0,zs[z]]) for x in range(0,len(xs))] for z in range(0,len(zs))])
B2 = np.array([[coilLoop.getB([xs[x],0,zs[z]]) for x in range(0,len(xs))] for z in range(0,len(zs))])
B3 = np.array([[coilLoop2.getB([xs[x],0,zs[z]]) for x in range(0,len(xs))] for z in range(0,len(zs))])


Bx = B[:,:,0]
By = B[:,:,1]
Bz = B[:,:,2]

B2x = B2[:,:,0]
B2y = B2[:,:,1]
B2z = B2[:,:,2]

B3x = B3[:,:,0]
B3y = B3[:,:,1]
B3z = B3[:,:,2]


fig = plt.figure(1)
AXS = [fig.add_subplot(2,2,i, axisbelow=True) for i in range(1,5)]
subPlot1,subPlot2,subPlot3,subPlot4 = AXS

subPlot1.contourf( X, Z, Bx,10,cmap=plt.cm.jet)
cp=subPlot1.contour( X, Z, Bx,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))

subPlot2.contourf( X, Z, By,10,cmap=plt.cm.jet)
cp=subPlot2.contour( X, Z, By,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))

subPlot3.contourf( X, Z, Bz,10,cmap=plt.cm.jet)
cp=subPlot3.contour( X, Z, Bz,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))


fig2 = plt.figure(2)
AXS2 = [fig2.add_subplot(2,2,i, axisbelow=True) for i in range(1,5)]
subPlot21,subPlot22,subPlot23,subPlot24 = AXS2

subPlot21.contourf( X, Z, B2x,10,cmap=plt.cm.jet)
cp=subPlot21.contour( X, Z, B2x,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))

subPlot22.contourf( X, Z, B2y,10,cmap=plt.cm.jet)
cp=subPlot22.contour( X, Z, B2y,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))

subPlot23.contourf( X, Z, B2z,10,cmap=plt.cm.jet)
cp=subPlot23.contour( X, Z, B2z,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))

fig3 = plt.figure(3)
AXS3 = [fig3.add_subplot(2,2,i, axisbelow=True) for i in range(1,5)]
subPlot31,subPlot32,subPlot33,subPlot34 = AXS3

subPlot21.contourf( X, Z, B2x,10,cmap=plt.cm.jet)
cp=subPlot31.contour( X, Z, B3x,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))

subPlot22.contourf( X, Z, B2y,10,cmap=plt.cm.jet)
cp=subPlot32.contour( X, Z, B3y,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))

subPlot23.contourf( X, Z, B2z,10,cmap=plt.cm.jet)
cp=subPlot33.contour( X, Z, B3z,30,colors=('k','k'),linestyles = ('-','-'),linewidths=(1,1))

plt.show()
```

If I use circular coils collection, I can get coils-effect. But lines collection doesn't work.
Is it a bug? or I just use magpylib incorrectly? Plz help me.

P.S This sample code is just sample. Actually, I want square 'coil's effect. So, I made some square's'. 
      I think just 4 of squares magnetic need sum, but, lots of square need 'Collection'.
      I also used previous magpylib, I could get correct result. However, this version(3.0.2) doesn't work maybe.

**Comment by OrtnerMichael** [comment]
hi @hotteock 

Thanks for bringing this to our attention. This is indeed a bug that you have found ! Unfortunately I cannot offer you a quick-fix, the arrangement of Lines in Collections goes a bit deeper - it might be a few days before we can bring an update.

What you can do right now is to use a simple loop instead. This will be (much) slower though.

Alternatively you can use the high-performance ``getBv`` top-level function and just put all Line segments inside. This will be very fast but it is not so easy to handle. It requires that you vectorize your code by hand (this is what Magpylib is doing for you automatically). I've taken the liberty to write you such a code :) - as you had to go through all this trouble.

I have renamed the Issue Tag into [Bug] and will work on fixing it in the near future.

```
import numpy as np
import matplotlib.pyplot as plt
import magpylib as magpy

# create a vector of vertex start adn end positions
start, end = [], []
for z in np.linspace(-2,2,20):
    start += [(-1,-1,z),(-1, 1,z),( 1, 1,z),( 1,-1,z)]
    end += [(-1,1,z),( 1, 1,z),( 1,-1,z),(-1,-1,z)]

# create a grid
ts = np.linspace(-3,3,30)
grid = np.array([(x,0,z) for z in ts for x in ts])

# tiling up to create the vectorized input
n_verts = 80
n_grid = 900
start = np.tile(start, (n_grid,1))
end = np.tile(end, (n_grid,1))
poso = np.repeat(grid, n_verts, axis=0)

# compute field on grid
B = magpy.getBv(
    source_type='Line',
    segment_start=start,
    segment_end=end,
    current=1,
    observer=poso)

# reshape long vector
B = B.reshape((n_grid, n_verts, 3))

# sum field over all Line segments
B = np.sum(B,axis=1)

# reshape grid and B for plotting
grid = grid.reshape((30,30,3))
B = B.reshape((30,30,3))

# field amplitude to make plot look nice
amp = np.linalg.norm(B, axis=2)

# display field in figure with matplotlib
plt.streamplot(grid[:,:,0], grid[:,:,2], B[:,:,0], B[:,:,2],
    density=2, color=np.log(amp), linewidth=1, cmap='autumn')

plt.tight_layout()
plt.show()

```

**Comment by OrtnerMichael** [comment]
hi @hotteock 

the bug is now fixed in the **v4_fixes** branch.

https://github.com/magpylib/magpylib/tree/v4_fixes

If you have time, please confirm that this solves your problem.

**Comment by OrtnerMichael** [comment]
This is now fixed in the development branch with #373

**Comment by hotteock** [comment]
Thanks to you the problem was solved.

I'm looking forward your project. Magpylib is so useful tools I thinked.
